alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/MonitorRunner.kt

```diff
@@ -66,80 +53,130 @@ import kotlinx.coroutines.launch
+        return this
+    }
+
+    fun registerScriptService(scriptService: ScriptService): MonitorRunner {
+        this.scriptService = scriptService
+        return this
+    }
+
+    fun registerSettings(settings: Settings): MonitorRunner {
+        this.settings = settings
+        return this
+    }
+
+    fun registerThreadPool(threadPool: ThreadPool): MonitorRunner {
+        this.threadPool = threadPool
+        return this
+    }
 
-    @Volatile private var hostDenyList = HOST_DENY_LIST.get(settings)
+    fun registerAlertIndices(alertIndices: AlertIndices): MonitorRunner {
+        this.alertIndices = alertIndices
+        return this
+    }
 
-    @Volatile private var destinationSettings = loadDestinationSettings(settings)
-    @Volatile private var destinationContextFactory = DestinationContextFactory(client, xContentRegistry, destinationSettings)
+    fun registerInputService(inputService: InputService): MonitorRunner {
+        this.inputService = inputService
+        return this
+    }
 
-    init {
+    fun registerTriggerService(triggerService: TriggerService): MonitorRunner {
+        this.triggerService = triggerService
+        return this
+    }
+
+    fun registerAlertService(alertService: AlertService): MonitorRunner {
+        this.alertService = alertService
+        return this
+    }
+
+    // Must be called after registerClusterService and registerSettings in AlertingPlugin
+    fun registerConsumers(): MonitorRunner {
+        retryPolicy = BackoffPolicy.constantBackoff(ALERT_BACKOFF_MILLIS.get(settings), ALERT_BACKOFF_COUNT.get(settings))
         clusterService.clusterSettings.addSettingsUpdateConsumer(ALERT_BACKOFF_MILLIS, ALERT_BACKOFF_COUNT) {
             millis, count -> retryPolicy = BackoffPolicy.constantBackoff(millis, count)
         }
+
+        moveAlertsRetryPolicy =
+            BackoffPolicy.exponentialBackoff(MOVE_ALERTS_BACKOFF_MILLIS.get(settings), MOVE_ALERTS_BACKOFF_COUNT.get(settings))
         clusterService.clusterSettings.addSettingsUpdateConsumer(MOVE_ALERTS_BACKOFF_MILLIS, MOVE_ALERTS_BACKOFF_COUNT) {
             millis, count -> moveAlertsRetryPolicy = BackoffPolicy.exponentialBackoff(millis, count)
         }
+
+        allowList = ALLOW_LIST.get(settings)
         clusterService.clusterSettings.addSettingsUpdateConsumer(ALLOW_LIST) {
             allowList = it
         }
+
+        // Host deny list is not a dynamic setting so no consumer is registered but the variable is set here
+        hostDenyList = HOST_DENY_LIST.get(settings)
+
+        return this
+    }
+
+    // To be safe, call this last as it depends on a number of other components being registered beforehand (client, settings, etc.)
+    fun registerDestinationSettings(): MonitorRunner {
+        destinationSettings = loadDestinationSettings(settings)
+        destinationContextFactory = DestinationContextFactory(client, xContentRegistry, destinationSettings)
+        return this
     }
```

Comment:
Should we use Builders or just a constructor, so we can enforce these variables are not null?

Reply:
I used a similar format to the runners in Index Management. The runner is a singleton now and should only be setup once in `AlertingPlugin.kt`. The class attributes are `lateinit`s which cannot be nullable. If `lateinit`s are accessed too early, they throw special exceptions. But since we register in the `createComponents()` method of `AlertingPlugin.kt` we should be fine.

Reply:
oh that is good to know, thanks